Unable to compile on windows Msys gcc 11.2.0/msvc 2022 pagination_range.h template overload

When compiling on either on MSYS gcc 11.2 or with Msvc make run into over load template issue with the code in pagination_range.h - this is a bug with the msvc / gcc compiler being unable to resolve which template function to use for various classes. 
The code was changed in version 1.21 and has not compiled since then.

```
typename StreamReader<T>::result_type GetNext() {
    if (current_ == page_.end()) {
      if (last_page_) return Status{};
      request_.set_page_token(std::move(token_));
      auto response = loader_(request_);
      if (!response.ok()) return std::move(response).status();
      token_ = ExtractPageToken(*response);
      if (token_.empty()) last_page_ = true;
      page_ = extractor_(*std::move(response));
      current_ = page_.begin();
      if (current_ == page_.end()) return Status{};
    }
    return std::move(*current_++);

template <typename U>
  static constexpr auto ExtractPageToken(U& u)
      -> decltype(std::move(*u.mutable_next_page_token())) {
    return std::move(*u.mutable_next_page_token());
  }
  template <typename U>
  static constexpr auto ExtractPageToken(U& u)
      -> decltype(std::move(u.next_page_token)) {
    return std::move(u.next_page_token);
  }
```
  
  For current environment - used Msys Pacman to get Grpc / nlohmann/json and libcurl  

  Built crc32c 
  C:/msys64/mingw64/bin/cmake.exe -H. -Bbuild-output -G "MinGW Makefiles" -DCMAKE_MAKE_PROGRAM=C:/msys64/mingw64/bin/mingw32-make.exe  -DCMAKE_C_COMPILER=C:/msys64/mingw64/bin/gcc.exe -DCMAKE_CXX_COMPILER=C:/msys64/mingw64/bin/g++.exe -DCMAKE_INSTALL_MESSAGE=NEVER -DCMAKE_BUILD_TYPE=Release -DBUILD_SHARED_LIBS=ON -DCMAKE_CXX_STANDARD=17 -DCRC32C_BUILD_TESTS=OFF  -DCRC32C_BUILD_BENCHMARKS=OFF  -DCRC32C_USE_GLOG=OFF 

built google-cloud-cpp via 
C:/msys64/mingw64/bin/cmake.exe -H. -Bbuild-output -G "MinGW Makefiles" -DCMAKE_MAKE_PROGRAM=C:/msys64/mingw64/bin/mingw32-make.exe  -DCMAKE_C_COMPILER=C:/msys64/mingw64/bin/gcc.exe -DCMAKE_CXX_COMPILER=C:/msys64/mingw64/bin/g++.exe -DCMAKE_INSTALL_MESSAGE=NEVER -DCMAKE_BUILD_TYPE=Release -DBUILD_SHARED_LIBS=ON -DCMAKE_CXX_STANDARD=17 -DBUILD_TESTING=OFF -DGOOGLE_CLOUD_CPP_ENABLE_EXAMPLES=OFF

And finally I was able to repro:

```
[932/1222] Building CXX object google/cloud/bigtable/CMakeFiles/google_cloud_cpp_bigtable.dir/admin/internal/bigtable_instance_admin_connection_impl.cc.obj
FAILED: google/cloud/bigtable/CMakeFiles/google_cloud_cpp_bigtable.dir/admin/internal/bigtable_instance_admin_connection_impl.cc.obj
C:\msys64\mingw64\bin\g++.exe -D_WIN32_WINNT=0x600 -Dgoogle_cloud_cpp_bigtable_EXPORTS -IC:/src/cpp-develop -isystem C:/src/cpp-develop/cmake-out/mingw64 -isystem C:/src/cpp-develop/cmake-out/mingw64/external/googleapis -O3 -DNDEBUG -Wall -Wextra -Wconversion -Wno-sign-conversion -Werror -std=gnu++11 -MD -MT google/cloud/bigtable/CMakeFiles/google_cloud_cpp_bigtable.dir/admin/internal/bigtable_instance_admin_connection_impl.cc.obj -MF google\cloud\bigtable\CMakeFiles\google_cloud_cpp_bigtable.dir\admin\internal\bigtable_instance_admin_connection_impl.cc.obj.d -o google/cloud/bigtable/CMakeFiles/google_cloud_cpp_bigtable.dir/admin/internal/bigtable_instance_admin_connection_impl.cc.obj -c C:/src/cpp-develop/google/cloud/bigtable/admin/internal/bigtable_instance_admin_connection_impl.cc
In file included from C:/src/cpp-develop/google/cloud/bigtable/admin/internal/bigtable_instance_admin_connection_impl.cc:25:
C:/src/cpp-develop/google/cloud/internal/pagination_range.h: In instantiation of 'typename google::cloud::v1_41_0::internal::StreamReader<T>::result_type google::cloud::v1_41_0::internal::PagedStreamReader<T, Request, Response>::GetNext() [with T = google::bigtable::admin::v2::AppProfile; Request = google::bigtable::admin::v2::ListAppProfilesRequest; Response = google::bigtable::admin::v2::ListAppProfilesResponse; typename google::cloud::v1_41_0::internal::StreamReader<T>::result_type = absl::lts_20211102::variant<google::cloud::v1_41_0::Status, google::bigtable::admin::v2::AppProfile>; google::cloud::v1_41_0::internal::StreamReader<T> = std::function<absl::lts_20211102::variant<google::cloud::v1_41_0::Status, google::bigtable::admin::v2::AppProfile>()>]':
C:/src/cpp-develop/google/cloud/internal/pagination_range.h:174:51:   required from 'Range google::cloud::v1_41_0::internal::MakePaginationRange(Request, Loader, Extractor) [with Range = google::cloud::v1_41_0::StreamRange<google::bigtable::admin::v2::AppProfile>; Request = google::bigtable::admin::v2::ListAppProfilesRequest; Loader = google::cloud::bigtable_admin_internal::v1_41_0::BigtableInstanceAdminConnectionImpl::ListAppProfiles(google::bigtable::admin::v2::ListAppProfilesRequest)::<lambda(const google::bigtable::admin::v2::ListAppProfilesRequest&)>; Extractor = google::cloud::bigtable_admin_internal::v1_41_0::BigtableInstanceAdminConnectionImpl::ListAppProfiles(google::bigtable::admin::v2::ListAppProfilesRequest)::<lambda(google::bigtable::admin::v2::ListAppProfilesResponse)>]'
C:/src/cpp-develop/google/cloud/bigtable/admin/internal/bigtable_instance_admin_connection_impl.cc:321:60:   required from here
C:/src/cpp-develop/google/cloud/internal/pagination_range.h:100:32: error: call of overloaded 'ExtractPageToken(google::bigtable::admin::v2::ListAppProfilesResponse&)' is ambiguous
  100 |       token_ = ExtractPageToken(*response);
      |                ~~~~~~~~~~~~~~~~^~~~~~~~~~~
C:/src/cpp-develop/google/cloud/internal/pagination_range.h:118:25: note: candidate: 'static constexpr decltype (std::move((* u.mutable_next_page_token()))) google::cloud::v1_41_0::internal::PagedStreamReader<T, Request, Response>::ExtractPageToken(U&) [with U = google::bigtable::admin::v2::ListAppProfilesResponse; T = google::bigtable::admin::v2::AppProfile; Request = google::bigtable::admin::v2::ListAppProfilesRequest; Response = google::bigtable::admin::v2::ListAppProfilesResponse; decltype (std::move((* u.mutable_next_page_token()))) = std::__cxx11::basic_string<char>&&]'
  118 |   static constexpr auto ExtractPageToken(U& u)
      |                         ^~~~~~~~~~~~~~~~
C:/src/cpp-develop/google/cloud/internal/pagination_range.h:123:25: note: candidate: 'static constexpr decltype (std::move(u.next_page_token)) google::cloud::v1_41_0::internal::PagedStreamReader<T, Request, Response>::ExtractPageToken(U&) [with U = google::bigtable::admin::v2::ListAppProfilesResponse; T = google::bigtable::admin::v2::AppProfile; Request = google::bigtable::admin::v2::ListAppProfilesRequest; Response = google::bigtable::admin::v2::ListAppProfilesResponse; decltype (std::move(u.next_page_token)) = const std::__cxx11::basic_string<char>& (google::bigtable::admin::v2::ListAppProfilesResponse::*&&)() const]'
  123 |   static constexpr auto ExtractPageToken(U& u)
      |                         ^~~~~~~~~~~~~~~~

```

--- a/google/cloud/internal/pagination_range.h
+++ b/google/cloud/internal/pagination_range.h
@@ -147,7 +147,7 @@ class PagedStreamReader {
    * For better or worse some versions of GCC called that ambiguous, even though
    * only one would work under SFINAE.
    */
-  static std::string ExtractPageToken(Response& u) {
+  static std::string ExtractPageToken(Response& r) {
     static_assert(
         HasMutableNextPageToken<Response>::value ||
             HasNextPageToken<Response>::value,
@@ -155,32 +155,35 @@ class PagedStreamReader {
     // This may seem expensive, but the value is known at compile type, the
     // optimizer should know what to do.
     if (HasMutableNextPageToken<Response>::value) {
-      return UsingMutableNextPageToken<Response>(u);
+      return UsingMutableNextPageToken<Response>(
+          HasMutableNextPageToken<Response>{}, r);
     }
-    return UsingNextPageToken<Response>(u);
+    return UsingNextPageToken<Response>(HasMutableNextPageToken<Response>{},
+                                        HasNextPageToken<Response>{}, r);
   }
 
   template <typename U>
-  static auto constexpr UsingMutableNextPageToken(U& u)
+  static auto constexpr UsingMutableNextPageToken(std::true_type, U& u)
       -> decltype(std::move(*u.mutable_next_page_token())) {
     return std::move(*u.mutable_next_page_token());
   }
   // This overload is unused, it is here just to compile when
   // HasMutableNextPageToken<Response>::value is `false`
-  template <typename U, typename... V>
-  static std::string UsingMutableNextPageToken(V&...) {
+  template <typename U>
+  static std::string UsingMutableNextPageToken(std::false_type, U&) {
     return {};
   }
 
   template <typename U>
-  static auto constexpr UsingNextPageToken(U& u)
+  static auto constexpr UsingNextPageToken(std::false_type, std::true_type,
+                                           U& u)
       -> decltype(std::move(u.next_page_token)) {
     return std::move(u.next_page_token);
   }
   // This overload is unused, it is here just to compile when
   // HasNextPageToken<Response>::value is `false`
-  template <typename... V>
-  static std::string UsingNextPageToken(V&...) {
+  template <typename U, typename D>
+  static std::string UsingNextPageToken(std::true_type, D, U&) {
     return {};
   }
 

--- a/google/cloud/internal/pagination_range.h
+++ b/google/cloud/internal/pagination_range.h
@@ -16,6 +16,7 @@
 #define GOOGLE_CLOUD_CPP_GOOGLE_CLOUD_INTERNAL_PAGINATION_RANGE_H
 
 #include "google/cloud/internal/invoke_result.h"
+#include "google/cloud/internal/type_traits.h"
 #include "google/cloud/status_or.h"
 #include "google/cloud/stream_range.h"
 #include "google/cloud/version.h"
@@ -107,22 +108,80 @@ class PagedStreamReader {
   }
 
  private:
+  template <typename U, typename AlwaysVoid = void>
+  struct HasMutableNextPageToken : public std::false_type {};
+  template <typename U>
+  struct HasMutableNextPageToken<
+      U,
+      void_t<decltype(std::move(*std::declval<U>().mutable_next_page_token()))>>
+      : public std::true_type {};
+
+  template <typename U, typename AlwaysVoid = void>
+  struct HasNextPageToken : public std::false_type {};
+  template <typename U>
+  struct HasNextPageToken<
+      U, void_t<decltype(std::move(std::declval<U>().next_page_token))>>
+      : public std::true_type {};
+
   /**
    * ExtractPageToken() extracts (i.e., "moves") the page token out of the
-   * given response object. This function is overloaded based on whether the
-   * response object has a `.mutable_next_page_token()` member function (e.g.,
-   * a protobuf), or a `.next_page_token` field (e.g., a regular struct such as
-   * is used in GCS).
+   * given response object. This function dispatches to the right function
+   * to extract the value using either a `.mutable_next_page_token()` member
+   * function (e.g., a protobuf), or a `.next_page_token` field (e.g., a regular
+   * struct such as is used in GCS).
+   *
+   * The code is more complicated than we would wish. It used to be (this
+   * comment uses C++17 for exposition purposes, we used the C++11 equivalent):
+   *
+   * @code
+   * template <typename U>
+   * static auto constexpr E(U& u) {
+   *   return std::move(*u.mutable_next_page_token());
+   * }
+   * template <typename U>
+   * static auto constexpr E(U& u) {
+   *   return std::move(u.next_page_token);
+   * }
+   * @endcode
+   *
+   * For better or worse some versions of GCC called that ambiguous, even though
+   * only one would work under SFINAE.
    */
+  static std::string ExtractPageToken(Response& u) {
+    static_assert(
+        HasMutableNextPageToken<Response>::value ||
+            HasNextPageToken<Response>::value,
+        "The Response type does not meet the requirements for PaginationRange");
+    // This may seem expensive, but the value is known at compile type, the
+    // optimizer should know what to do.
+    if (HasMutableNextPageToken<Response>::value) {
+      return UsingMutableNextPageToken<Response>(u);
+    }
+    return UsingNextPageToken<Response>(u);
+  }
+
   template <typename U>
-  static constexpr auto ExtractPageToken(U& u)
+  static auto constexpr UsingMutableNextPageToken(U& u)
       -> decltype(std::move(*u.mutable_next_page_token())) {
     return std::move(*u.mutable_next_page_token());
   }
+  // This overload is unused, it is here just to compile when
+  // HasMutableNextPageToken<Response>::value is `false`
+  template <typename U, typename... V>
+  static std::string UsingMutableNextPageToken(V&...) {
+    return {};
+  }
+
   template <typename U>
-  static constexpr auto ExtractPageToken(U& u)
+  static auto constexpr UsingNextPageToken(U& u)
       -> decltype(std::move(u.next_page_token)) {
     return std::move(u.next_page_token);
+  }
+  // This overload is unused, it is here just to compile when
+  // HasNextPageToken<Response>::value is `false`
+  template <typename... V>
+  static std::string UsingNextPageToken(V&...) {
+    return {};
   }
 
   Request request_;

--- a/google/cloud/internal/pagination_range_test.cc
+++ b/google/cloud/internal/pagination_range_test.cc
@@ -51,6 +51,7 @@ struct ProtoResponse {
   std::vector<Item> testonly_items;
   std::string testonly_page_token;
   std::string* mutable_next_page_token() { return &testonly_page_token; }
+  std::string const& next_page_token() const { return testonly_page_token; }
 
   // Used for setting the token in tests, but it's not used in the real code.
   void testonly_set_page_token(std::string s) {

